File: rust/device.rs
//! Generic device functionality which is used by all bricks and bricklets.

use crate::{
    base58::*,
    byte_converter::FromByteSlice,
    converting_callback_receiver::ConvertingCallbackReceiver,
    converting_receiver::{BrickletError, BrickletRecvTimeoutError, ConvertingReceiver},
    ip_connection::{IpConnection, Request, SocketThreadRequest},
    low_level_traits::*,
};
use std::sync::{
    mpsc::{channel, Sender},
    Arc, Mutex,
};

#[derive(Debug, Copy, Clone, PartialEq)]
pub(crate) enum ResponseExpectedFlag {
    InvalidFunctionId,
    False,
    True,
    AlwaysTrue,
}

impl From<bool> for ResponseExpectedFlag {
    fn from(b: bool) -> Self {
        if b {
            ResponseExpectedFlag::True
        } else {
            ResponseExpectedFlag::False
        }
    }
}

#[derive(Clone)]
pub(crate) struct Device {
    pub api_version: [u16; 3],
    pub response_expected: [ResponseExpectedFlag; 256],
    pub internal_uid: u32,
    pub req_tx: Sender<SocketThreadRequest>,
    pub high_level_locks: Vec<Arc<Mutex<()>>>,
}

/// This error is returned if the response expected status was queried for an unknown function.
#[derive(Debug, Copy, Clone)]
pub struct GetResponseExpectedError(u8);

impl std::error::Error for GetResponseExpectedError {}

impl std::fmt::Display for GetResponseExpectedError {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        write!(f, "Can not get response expected: Invalid function id {}", self.0)
    }
}

/// This error is returned if the response expected status of a function could not be changed.
#[derive(Debug, Copy, Clone)]
pub enum SetResponseExpectedError {
    /// The function id was unknown. Maybe the wrong UID was used?
    InvalidFunctionId(u8),
    /// This function must always respond, as the response contains result data.
    IsAlwaysTrue(u8),
}

impl std::error::Error for SetResponseExpectedError {}

impl std::fmt::Display for SetResponseExpectedError {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        match self {
            SetResponseExpectedError::InvalidFunctionId(fid) => write!(f, "Can not set response expected: Invalid function id {}", fid),
            SetResponseExpectedError::IsAlwaysTrue(_fid) => write!(f, "Can not set response expected: function always responds."),
        }
    }
}

impl Device {
    pub(crate) fn new(api_version: [u16; 3], uid: &str, ip_connection: &IpConnection, high_level_function_count: u8) -> Device {
        Device {
            api_version,
            internal_uid: uid.base58_to_u32().unwrap(),
            req_tx: ip_connection.req.socket_thread_tx.clone(),
            response_expected: [ResponseExpectedFlag::InvalidFunctionId; 256],
            high_level_locks: vec![Arc::new(Mutex::new(())); high_level_function_count as usize],
        }
    }

    pub(crate) fn get_response_expected(&self, function_id: u8) -> Result<bool, GetResponseExpectedError> {
        match self.response_expected[function_id as usize] {
            ResponseExpectedFlag::False => Ok(false),
            ResponseExpectedFlag::True => Ok(true),
            ResponseExpectedFlag::AlwaysTrue => Ok(true),
            ResponseExpectedFlag::InvalidFunctionId => Err(GetResponseExpectedError(function_id)),
        }
    }

    pub(crate) fn set_response_expected(&mut self, function_id: u8, response_expected: bool) -> Result<(), SetResponseExpectedError> {
        if self.response_expected[function_id as usize] == ResponseExpectedFlag::AlwaysTrue {
            Err(SetResponseExpectedError::IsAlwaysTrue(function_id))
        } else if self.response_expected[function_id as usize] == ResponseExpectedFlag::InvalidFunctionId {
            Err(SetResponseExpectedError::InvalidFunctionId(function_id))
        } else {
            self.response_expected[function_id as usize] = ResponseExpectedFlag::from(response_expected);
            Ok(())
        }
    }

    pub(crate) fn set_response_expected_all(&mut self, response_expected: bool) {
        for resp_exp in self.response_expected.iter_mut() {
            if *resp_exp == ResponseExpectedFlag::True || *resp_exp == ResponseExpectedFlag::False {
                *resp_exp = ResponseExpectedFlag::from(response_expected);
            }
        }
    }

    pub(crate) fn set<T: FromByteSlice>(&self, function_id: u8, payload: Vec<u8>) -> ConvertingReceiver<T> {
        let (sent_tx, sent_rx) = channel();
        if self.response_expected[function_id as usize] == ResponseExpectedFlag::False {
            let (tx, rx) = channel();
            self.req_tx
                .send(SocketThreadRequest::Request(
                    Request::Set { uid: self.internal_uid, function_id, payload, response_sender: None },
                    sent_tx,
                ))
                .expect("The socket thread queue was disconnected from the ip connection. This is a bug in the rust bindings.");
            let timeout = sent_rx.recv().expect("The sent queue was dropped. This is a bug in the rust bindings.");
            let _ = tx.send(Err(BrickletError::SuccessButResponseExpectedIsDisabled));
            ConvertingReceiver::new(rx, timeout)
        } else {
            let (tx, rx) = channel();
            self.req_tx
                .send(SocketThreadRequest::Request(
                    Request::Set { uid: self.internal_uid, function_id, payload, response_sender: Some(tx) },
                    sent_tx,
                ))
                .expect("The socket thread queue was disconnected from the ip connection. This is a bug in the rust bindings.");
            let timeout = sent_rx.recv().expect("The sent queue was dropped. This is a bug in the rust bindings.");
            ConvertingReceiver::new(rx, timeout)
        }
    }

    pub(crate) fn get_callback_receiver<T: FromByteSlice>(&self, function_id: u8) -> ConvertingCallbackReceiver<T> {
        let (tx, rx) = channel();
        let (sent_tx, sent_rx) = channel();
        self.req_tx
            .send(SocketThreadRequest::Request(
                Request::RegisterCallback { uid: self.internal_uid, function_id, response_sender: tx },
                sent_tx,
            ))
            .expect("The socket thread queue was disconnected from the ip connection. This is a bug in the rust bindings.");
        sent_rx.recv().expect("The sent queue was dropped. This is a bug in the rust bindings.");
        ConvertingCallbackReceiver::new(rx)
    }

    pub(crate) fn get<T: FromByteSlice>(&self, function_id: u8, payload: Vec<u8>) -> ConvertingReceiver<T> {
        let (tx, rx) = channel();
        let (sent_tx, sent_rx) = channel();
        self.req_tx
            .send(SocketThreadRequest::Request(Request::Get { uid: self.internal_uid, function_id, payload, response_sender: tx }, sent_tx))
            .expect("The socket thread queue was disconnected from the ip connection. This is a bug in the rust bindings.");
        let timeout = sent_rx.recv().expect("The sent queue was dropped. This is a bug in the rust bindings.");
        ConvertingReceiver::new(rx, timeout)
    }

    pub(crate) fn set_high_level<
        PayloadT,
        OutputT,
        LlwT: LowLevelWrite<OutputT>,
        ClosureT: FnMut(usize, usize, &[PayloadT]) -> Result<LlwT, BrickletRecvTimeoutError>,
    >(
        &self,
        high_level_function_idx: u8,
        payload: &[PayloadT],
        max_payload_len: usize,
        chunk_len: usize,
        low_level_closure: &mut ClosureT,
    ) -> Result<(usize, OutputT), BrickletRecvTimeoutError> {
        if payload.len() > max_payload_len {
            return Err(BrickletRecvTimeoutError::InvalidParameter);
        }

        let length = payload.len();

        let mut chunk_offset = 0;
        {
            let _lock_guard = self.high_level_locks[high_level_function_idx as usize].lock().unwrap();
            if length == 0 {
                match low_level_closure(length, chunk_offset, &[]) {
                    Ok(low_level_result) => return Ok((low_level_result.ll_message_written(), low_level_result.get_result())),
                    Err(e) => return Err(e),
                }
            }
            let mut written_sum = 0;
            loop {
                match low_level_closure(length, chunk_offset, &payload[chunk_offset..std::cmp::min(chunk_offset + chunk_len, length)]) {
                    Ok(low_level_result) => {
                        let written = low_level_result.ll_message_written();
                        let output = low_level_result.get_result();
                        written_sum += written;
                        if written < chunk_len {
                            return Ok((written_sum, output));
                        }
                        chunk_offset += chunk_len;
                        if chunk_offset >= length {
                            return Ok((written_sum, output));
                        }
                    }
                    Err(e) => return Err(e),
                }
            }
        }
    }

    pub(crate) fn get_high_level<
        PayloadT: Default + Clone + Copy,
        OutputT,
        LlrT: LowLevelRead<PayloadT, OutputT>,
        ClosureT: FnMut() -> Result<LlrT, BrickletRecvTimeoutError>,
    >(
        &self,
        high_level_function_idx: u8,
        low_level_closure: &mut ClosureT,
    ) -> Result<(Vec<PayloadT>, OutputT), BrickletRecvTimeoutError> {
        let mut chunk_offset = 0;
        {
            let _lock_guard = self.high_level_locks[high_level_function_idx as usize].lock().unwrap();
            let mut result = low_level_closure()?;
            let mut out_of_sync = result.ll_message_chunk_offset() != 0;
            let message_length = result.ll_message_length();

            if !out_of_sync {
                let mut buf = vec![PayloadT::default(); message_length];
                let first_read_length = std::cmp::min(result.ll_message_chunk_data().len(), message_length - chunk_offset);
                buf[chunk_offset..chunk_offset + first_read_length].copy_from_slice(&result.ll_message_chunk_data()[0..first_read_length]);
                chunk_offset += first_read_length;
                while chunk_offset < message_length {
                    result = low_level_closure()?;
                    out_of_sync = result.ll_message_chunk_offset() != chunk_offset || result.ll_message_length() != message_length;
                    if out_of_sync {
                        break;
                    }

                    let read_length = std::cmp::min(result.ll_message_chunk_data().len(), message_length - chunk_offset);
                    buf[chunk_offset..chunk_offset + read_length].copy_from_slice(&result.ll_message_chunk_data()[0..read_length]);
                    chunk_offset += read_length;
                }
                if !out_of_sync {
                    return Ok((buf, result.get_result()));
                }
            }

            assert!(out_of_sync);
            while chunk_offset + result.ll_message_chunk_data().len() < message_length {
                result = low_level_closure()?;
            }
            Err(BrickletRecvTimeoutError::MalformedPacket)
        }
    }
}

File: go/device.go
/* Copyright (C) 2018-2020 Erik Fleckstein <erik@tinkerforge.com>
 *
 * Redistribution and use in source and binary forms of this file,
 * with or without modification, are permitted. See the Creative
 * Commons Zero (CC0 1.0) License for more details.
 */

package internal

import (
	"bytes"
	"encoding/binary"
	"fmt"
	"sync"
)

type ResponseExpectedFlag = uint8

const (
	ResponseExpectedFlagInvalidFunctionID ResponseExpectedFlag = iota
	ResponseExpectedFlagFalse
	ResponseExpectedFlagTrue
	ResponseExpectedFlagAlwaysTrue
)

type deviceIdentifierCheck = uint8

const (
	deviceIdentifierCheckPending deviceIdentifierCheck = iota
	deviceIdentifierCheckMatch
	deviceIdentifierCheckMismatch
)

type Device struct {
	apiVersion             [3]uint8
	ResponseExpected       [256]ResponseExpectedFlag
	internalUID            uint32
	regTX                  chan<- Request
	callbackRegTX          chan<- CallbackRegistration
	callbackDeregTX        chan<- CallbackDeregistration
	highLevelLocks         []sync.Mutex
	initialized            bool
	deviceIdentifierCheck  deviceIdentifierCheck
	deviceIdentifierMutex  sync.Mutex
	deviceIdentifier       uint16
	deviceDisplayName      string
	wrongDeviceDisplayName string
	UID                    string
}

func NewDevice(apiVersion [3]uint8, uid string, ipcon *IPConnection, highLevelFunctionCount uint8, deviceIdentifier uint16, deviceDisplayName string) (Device, error) {
	internalUID, err := Base58ToU32(uid)
	if err != nil {
		result := Device{}
		result.initialized = false
		return result, err
	}
	return Device{
		apiVersion,
		[256]ResponseExpectedFlag{},
		internalUID,
		ipcon.Req,
		ipcon.CallbackReg,
		ipcon.CallbackDereg,
		make([]sync.Mutex, highLevelFunctionCount),
		true,
		deviceIdentifierCheckPending,
		sync.Mutex{},
		deviceIdentifier,
		deviceDisplayName,
		"",
		uid}, nil
}

func (device *Device) GetAPIVersion() [3]uint8 {
	return device.apiVersion
}

func (device *Device) GetResponseExpected(functionID uint8) (bool, error) {
	switch device.ResponseExpected[functionID] {
	case ResponseExpectedFlagAlwaysTrue:
		return true, nil
	case ResponseExpectedFlagTrue:
		return true, nil
	case ResponseExpectedFlagFalse:
		return false, nil
	default:
		return false, fmt.Errorf("unknown function with ID %d", functionID)
	}
}

func (device *Device) SetResponseExpected(functionID uint8, responseExpected bool) error {
	switch device.ResponseExpected[functionID] {
	case ResponseExpectedFlagAlwaysTrue:
		return fmt.Errorf("response expected for function with ID %d is always true", functionID)
	case ResponseExpectedFlagInvalidFunctionID:
		return fmt.Errorf("unknown function with ID %d", functionID)
	default:
		if responseExpected {
			device.ResponseExpected[functionID] = ResponseExpectedFlagTrue
		} else {
			device.ResponseExpected[functionID] = ResponseExpectedFlagFalse
		}
	}
	return nil
}

func (device *Device) SetResponseExpectedAll(responseExpected bool) {
	for i, respExp := range device.ResponseExpected {
		if respExp == ResponseExpectedFlagTrue || respExp == ResponseExpectedFlagFalse {
			if responseExpected {
				device.ResponseExpected[i] = ResponseExpectedFlagTrue
			} else {
				device.ResponseExpected[i] = ResponseExpectedFlagFalse
			}
		}
	}
}

func (device *Device) Set(functionID uint8, payload []byte) ([]byte, error) {
	if !device.initialized {
		return nil, fmt.Errorf("device is not initialized")
	}
	err := device.checkDeviceIdentifier()
	if err != nil {
		return nil, err
	}

	responseExpected := !(device.ResponseExpected[functionID] == ResponseExpectedFlagFalse)

	if responseExpected {
		return device.Get(functionID, payload)
	}

	header := PacketHeader{
		device.internalUID,
		uint8(len(payload) + PacketHeaderSize),
		functionID,
		0,
		false,
		0}

	bytes := append(header.ToLeBytes(), payload[:]...)
	device.regTX <- Request{bytes[:], nil, nil}
	return nil, nil
}

func (device *Device) Get(functionID uint8, payload []byte) ([]byte, error) {
	if !device.initialized {
		return nil, fmt.Errorf("device is not initialized")
	}

	err := device.checkDeviceIdentifier()
	if err != nil {
		return nil, err
	}

	header := PacketHeader{
		device.internalUID,
		uint8(len(payload) + PacketHeaderSize),
		functionID,
		0,
		true,
		0}
	bytes := append(header.ToLeBytes(), payload[:]...)

	return ReqWithTimeout(bytes, device.regTX)
}

type LowLevelWriteResult struct {
	Written uint64
	Result  []byte
}

func (device *Device) SetHighLevel(lowLevelClosure func(uint64, uint64, []byte) (LowLevelWriteResult, error), highLevelFunctionIdx uint8, elementSizeInBit uint64, chunkLenInBit uint64, payload []byte) (LowLevelWriteResult, error) {
	if !device.initialized {
		return LowLevelWriteResult{}, fmt.Errorf("device is not initialized")
	}
	length := uint64(len(payload)) * 8 / elementSizeInBit
	chunkOffset := uint64(0) * 8 / elementSizeInBit

	device.highLevelLocks[highLevelFunctionIdx].Lock()
	defer device.highLevelLocks[highLevelFunctionIdx].Unlock()

	if length == 0 {
		return lowLevelClosure(length, chunkOffset, []byte{})
	}
	writtenSum := uint64(0)
	for {
		toWrite := MinU(uint64(chunkOffset+chunkLenInBit), uint64(length))
		result, err := lowLevelClosure(length, chunkOffset, payload[chunkOffset/8:toWrite/8])
		if err != nil {
			return result, err
		}
		written := result.Written * 8
		writtenSum += written
		if written < chunkLenInBit {
			result.Written = writtenSum / 8
			return result, nil
		}
		chunkOffset += chunkLenInBit
		if chunkOffset >= length {
			result.Written = length / 8
			return result, nil
		}
	}
}

type LowLevelResult struct {
	Length      uint64
	ChunkOffset uint64
	ChunkData   []byte
	Result      []byte
}

func (device *Device) GetHighLevel(lowLevelClosure func() (LowLevelResult, error), highLevelFunctionIdx uint8, elementSizeInBit uint64) ([]byte, []byte, error) {
	if !device.initialized {
		return nil, nil, fmt.Errorf("device is not initialized")
	}
	chunkOffset := uint64(0) * elementSizeInBit
	device.highLevelLocks[highLevelFunctionIdx].Lock()
	defer device.highLevelLocks[highLevelFunctionIdx].Unlock()

	result, err := lowLevelClosure()
	if err != nil {
		return nil, nil, err
	}
	outOfSync := result.ChunkOffset != 0
	messageLength := result.Length * elementSizeInBit

	if !outOfSync {
		buf := make([]byte, 0, messageLength/8)
		firstReadLen := MinU(uint64(len(result.ChunkData)*8), uint64(messageLength-chunkOffset))
		buf = append(buf, result.ChunkData[0:firstReadLen/8]...)
		chunkOffset += firstReadLen
		for chunkOffset < messageLength {
			result, err = lowLevelClosure()
			if err != nil {
				return nil, nil, err
			}
			outOfSync = result.ChunkOffset*elementSizeInBit != chunkOffset || result.Length*elementSizeInBit != messageLength
			if outOfSync {
				break
			}
			readLen := MinU(uint64(len(result.ChunkData)*8), uint64(messageLength-chunkOffset))
			buf = append(buf, result.ChunkData[0:readLen/8]...)
			chunkOffset += readLen
		}
		if !outOfSync {
			return buf, result.Result, nil
		}
	}

	chunkOffset = result.ChunkOffset
	for chunkOffset+uint64(len(result.ChunkData)*8) < messageLength {
		chunkOffset += uint64(len(result.ChunkData) * 8)
		result, _ = lowLevelClosure()
	}
	return nil, nil, fmt.Errorf("stream is out of sync, please retry")
}

func (device *Device) checkDeviceIdentifier() error {
	if device.deviceIdentifierCheck == deviceIdentifierCheckMatch {
		return nil
	}

	device.deviceIdentifierMutex.Lock()
	defer device.deviceIdentifierMutex.Unlock()

	if device.deviceIdentifierCheck == deviceIdentifierCheckPending {
		reqHeader := PacketHeader{
			device.internalUID,
			uint8(PacketHeaderSize),
			255,
			0,
			true,
			0}
		reqBytes := reqHeader.ToLeBytes()

		resultBytes, err := ReqWithTimeout(reqBytes, device.regTX)
		if err != nil {
			return err
		}
		if len(resultBytes) < 8 {
			return fmt.Errorf("get identity returned packet of unexpected size %d", len(resultBytes))
		}
		var header PacketHeader

		header.FillFromBytes(resultBytes)
		if header.ErrorCode != 0 {
			return DeviceError(header.ErrorCode)
		}

		if header.Length < 33 {
			return fmt.Errorf("get identity returned packet of unexpected size %d", header.Length)
		}

		resultBuf := bytes.NewBuffer(resultBytes[8:])
		resultBuf.Next(8 + 8 + 1 + 3 + 3)
		var deviceIdentifier uint16
		binary.Read(resultBuf, binary.LittleEndian, &deviceIdentifier)
		if deviceIdentifier != device.deviceIdentifier {
			device.deviceIdentifierCheck = deviceIdentifierCheckMismatch
			device.wrongDeviceDisplayName = getDeviceDisplayName(deviceIdentifier)
		}
	}

	if device.deviceIdentifierCheck == deviceIdentifierCheckMismatch {
		return fmt.Errorf("UID %v belongs to a %v instead of the expected %v", device.UID, device.wrongDeviceDisplayName, device.deviceDisplayName)
	}
	return nil
}

func (device *Device) RegisterCallback(functionID uint8, fn func([]byte)) uint64 {
	wrapper := func(byteSlice []byte) {
		err := device.checkDeviceIdentifier()
		if err != nil {
			return
		}

		fn(byteSlice)
	}
	idChan := make(chan uint64, ChannelSize)
	device.callbackRegTX <- CallbackRegistration{device.internalUID, functionID, wrapper, idChan}
	return <-idChan
}

func (device *Device) DeregisterCallback(functionID uint8, registrationID uint64) {
	device.callbackDeregTX <- CallbackDeregistration{device.internalUID, functionID, registrationID}
}
